File: LinkedList/234-Palindrome-Linked-List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        #find the mid node
        s = f = head
        while f and f.next:
            s = s.next
            f = f.next.next
        #reverse the 2nd half
        N = None
        while s:
            tmp = s
            s = s.next
            tmp.next = N
            N = tmp
        #N is now the new head of the 2nd 
        p = head
        while N:
            if N.val != p.val:
                return False
            N = N.next
            p = p.next
        return True
        

File: LinkedList/234-Palindrome-Linked-List.java
/*
https://leetcode.com/problems/palindrome-linked-list/
ref: https://leetcode.com/discuss/78152/java-easy-to-understand
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isPalindrome(ListNode head) {
        //find the mid node
        ListNode s = head;
        ListNode f = head;
        ListNode p = head;
        while (f!=null && f.next!=null){
            s = s.next;
            f = f.next.next;
        }
        //reverse the 2nd half
        ListNode N = new ListNode(0);
        ListNode Terminate = N;
        while (s!=null){
            ListNode tmp = s;
            s = s.next;
            tmp.next = N;
            N = tmp;
        }
        //N is not the new head of 2nd half
        while (N!=Terminate){
            if (N.val!=p.val){
                return false;
            }
            N = N.next;
            p = p.next;
        }
        return true;
    }
}
